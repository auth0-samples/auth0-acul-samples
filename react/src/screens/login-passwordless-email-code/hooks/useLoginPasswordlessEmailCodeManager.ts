import {
  ScreenMembersOnLoginPasswordlessEmailCode,
  useLoginPasswordlessEmailCode,
} from "@auth0/auth0-acul-react";

import { executeSafely } from "@/utils/helpers/executeSafely";

/**
 * Handles the login process using a email and a code.
 * Optionally includes a CAPTCHA value if required.
 *
 * @param captcha - (Optional) The CAPTCHA value if required.
 * @param code - Code sent to the user's email.
 * @returns A promise that resolves when the login process is complete.
 */
export const useLoginPasswordlessEmailCodeManager = () => {
  const loginPasswordlessEmailCode = useLoginPasswordlessEmailCode();
  const { transaction, screen } = loginPasswordlessEmailCode;

  const { texts, captchaImage, data, links } = screen;

  const handleSubmitEmailCode = async (
    code: string,
    captcha?: string
  ): Promise<void> => {
    const options: { code: string; captcha?: string } = {
      code: code?.trim() || "",
    };
    if (screen.isCaptchaAvailable && captcha?.trim()) {
      options.captcha = captcha.trim();
    }
    executeSafely(
      `Submit Email Code with options: ${JSON.stringify(options)}`,
      () => loginPasswordlessEmailCode.submitCode(options)
    );
  };

  const handleResendEmailCode = async (payload?: never): Promise<void> => {
    executeSafely(
      `Resent Email Code with Custom options: ${JSON.stringify(payload)}`,
      () => loginPasswordlessEmailCode.resendCode(payload)
    );
  };

  return {
    loginPasswordlessEmailCode,
    handleSubmitEmailCode,
    handleResendEmailCode,
    texts: (texts || {}) as ScreenMembersOnLoginPasswordlessEmailCode["texts"],
    isCaptchaAvailable: screen.isCaptchaAvailable === true,
    errors: transaction.errors || [],
    captchaImage,
    data,
    links,
  };
};
