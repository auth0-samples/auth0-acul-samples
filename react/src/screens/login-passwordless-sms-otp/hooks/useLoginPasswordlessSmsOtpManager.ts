import {
  ScreenMembersOnLoginPasswordlessSmsOtp,
  useLoginPasswordlessSmsOtp,
} from "@auth0/auth0-acul-react";

import { executeSafely } from "@/utils/helpers/executeSafely";

/**
 * Handles the login process using a username or phone number and a One time password.
 * Optionally includes a CAPTCHA value if required.
 *
 * @param captcha - (Optional) The CAPTCHA value if required.
 * @param code - One time password (OTP) sent to the user's phone.
 * @param username - Username or phone number of the user.
 * @returns A promise that resolves when the login process is complete.
 */
export const useLoginPasswordlessSmsOtpManager = () => {
  const loginPasswordlessSmsOtp = useLoginPasswordlessSmsOtp();
  const { transaction, screen } = loginPasswordlessSmsOtp;

  const { texts, captchaImage, data, links } = screen;

  const handleSubmitOTP = async (
    username: string,
    code: string,
    captcha?: string
  ): Promise<void> => {
    const options: { username: string; code: string; captcha?: string } = {
      username: username?.trim() || "",
      code: code?.trim() || "",
    };
    if (screen.isCaptchaAvailable && captcha?.trim()) {
      options.captcha = captcha.trim();
    }
    executeSafely(
      `Submit OTP with options: ${JSON.stringify(options)}`,
      () => () => loginPasswordlessSmsOtp.submitOTP(options)
    );
  };

  const handleResendOTP = async (payload?: never): Promise<void> => {
    executeSafely(
      `Resent OTP with Custom options: ${JSON.stringify(payload)}`,
      () => loginPasswordlessSmsOtp.resendOTP(payload)
    );
  };

  return {
    loginPasswordlessSmsOtp,
    handleSubmitOTP,
    handleResendOTP,
    texts: (texts || {}) as ScreenMembersOnLoginPasswordlessSmsOtp["texts"],
    isCaptchaAvailable: screen.isCaptchaAvailable === true,
    errors: transaction.errors || [],
    captchaImage,
    data,
    links,
  };
};
