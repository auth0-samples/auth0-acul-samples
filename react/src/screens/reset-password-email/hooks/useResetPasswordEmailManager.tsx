import {
  ScreenMembersOnResetPasswordEmail,
  useResetPasswordEmail,
} from "@auth0/auth0-acul-react";

import { executeSafely } from "@/utils/helpers/executeSafely";

/**
 * Handles successfully sent email screen for password reset process
 * along with resend option
 *
 * @returns A promise that resolves when the email resend process is complete.
 */
export const useResetPasswordEmailManager = () => {
  const resetPasswordEmail = useResetPasswordEmail();
  const { screen, transaction } = resetPasswordEmail;

  const { texts, data } = screen;

  const handleResendEmail = async (): Promise<void> => {
    executeSafely(`Resend email for password reset`, () =>
      resetPasswordEmail.resendEmail()
    );
  };

  return {
    resetPasswordEmail,
    handleResendEmail,
    texts: (texts || {}) as ScreenMembersOnResetPasswordEmail["texts"],
    errors: transaction.errors || [],
    data: data || {},
  };
};
