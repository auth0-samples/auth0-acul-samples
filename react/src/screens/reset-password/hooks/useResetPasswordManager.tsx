import {
  resetPassword,
  ScreenMembersOnResetPassword,
  useResetPassword,
  useScreen,
  useTransaction,
} from "@auth0/auth0-acul-react/reset-password";

import { executeSafely } from "@/utils/helpers/executeSafely";

/**
 * Handles the password reset process
 *
 *
 * @param new_password - New password string.
 * @param confirm_password - New password string.
 * @returns A promise that resolves when the password reset process is complete.
 */
export const useResetPasswordManager = () => {
  const { texts } = useScreen();

  const handleSubmitPassword = async (
    new_password: string,
    confirm_password: string
  ): Promise<void> => {
    const options: { "password-reset": string; "re-enter-password": string } = {
      "password-reset": new_password?.trim() || "",
      "re-enter-password": confirm_password?.trim() || "",
    };
    executeSafely(
      `Submit password reset with options: ${JSON.stringify(options)}`,
      () => resetPassword(options)
    );
  };

  return {
    resetPassword: useResetPassword(),
    handleSubmitPassword,
    texts: (texts || {}) as ScreenMembersOnResetPassword["texts"],
    errors: useTransaction().errors || [],
  };
};
