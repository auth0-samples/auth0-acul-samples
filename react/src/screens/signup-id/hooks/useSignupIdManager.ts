import type {
  FederatedSignupOptions,
  SignupOptions,
} from "@auth0/auth0-acul-js/signup-id";
import { useSignupId } from "@auth0/auth0-acul-react/signup-id";
import { useScreen, useTransaction } from "@auth0/auth0-acul-react/signup-id";

import { executeSafely } from "@/utils/helpers/executeSafely";

export const useSignupIdManager = () => {
  const signupId = useSignupId();

  const screen = useScreen();
  const transaction = useTransaction();
  const { alternateConnections } = transaction;

  const { isCaptchaAvailable, texts, loginLink, captchaImage } = screen;

  const handleSignup = async (payload: SignupOptions): Promise<void> => {
    executeSafely(`Signup with options: ${JSON.stringify(payload)}`, () =>
      signupId.signup(payload)
    );
  };

  const handleFederatedSignup = async (payload: FederatedSignupOptions) => {
    executeSafely(
      `Federated signup with connection: ${payload.connection}`,
      () => signupId.federatedSignup(payload)
    );
  };

  const handlePickCountryCode = async (): Promise<void> => {
    executeSafely(`Pick country code`, () => signupId.pickCountryCode());
  };

  return {
    signupId,
    handleSignup,
    handleFederatedSignup,
    handlePickCountryCode,
    texts,
    isCaptchaAvailable,
    loginLink,
    alternateConnections,
    captchaImage,
    errors: transaction.errors,
  };
};
