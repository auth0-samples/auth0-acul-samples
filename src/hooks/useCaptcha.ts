import { useState, useCallback } from "react";
import Captcha from "@/common/Captcha";

export const useCaptcha = (
  captcha: any,
  label?: string,
  errorText?: string,
) => {
  const [isValid, setIsValid] = useState(!captcha);
  const [value, setValue] = useState<string | undefined>();
  const [error, setError] = useState("");

  const handleValidationChange = useCallback(
    (valid: boolean, val?: string, err?: string) => {
      setIsValid(valid);
      setValue(val);
      setError(err || "");
    },
    [],
  );

  const validate = useCallback(() => {
    if (!captcha) return true;
    if (!isValid) {
      setError(errorText || "");
      return false;
    }
    setError("");
    return true;
  }, [captcha, isValid, errorText]);

  const clearError = useCallback(() => setError(""), []);

  const CaptchaComponent = captcha
    ? Captcha({
        captcha,
        label,
        onValidationChange: handleValidationChange,
      })
    : null;

  return {
    Captcha: CaptchaComponent,
    isValid,
    value,
    error,
    validate,
    clearError,
  };
};
