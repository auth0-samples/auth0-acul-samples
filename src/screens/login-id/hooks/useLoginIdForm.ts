import { useRef } from "react";

export const useLoginIdForm = () => {
  const identifierRef = useRef<HTMLInputElement>(
    null,
  ) as React.RefObject<HTMLInputElement>;
  const captchaRef = useRef<HTMLInputElement>(
    null,
  ) as React.RefObject<HTMLInputElement>;

  const getFormValues = () => ({
    identifier: identifierRef.current?.value ?? "",
    captcha: captchaRef.current?.value ?? "",
  });

  return {
    identifierRef,
    captchaRef,
    getFormValues,
  };
};
