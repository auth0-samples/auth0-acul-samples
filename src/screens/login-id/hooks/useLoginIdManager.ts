import { useState } from "react";
import LoginIdInstance from "@auth0/auth0-acul-js/login-id";
import { executeSafely } from "@/utils/executeSafely";

export const useLoginIdManager = () => {
  const [loginIdInstance] = useState(() => new LoginIdInstance());

  const handleLoginId = (loginId: string, captcha?: string): void => {
    const options = {
      username: loginId,
      captcha: loginIdInstance.screen?.captcha ? captcha : undefined,
    };
    executeSafely(`LoginId with options: ${JSON.stringify(options)}`, () =>
      loginIdInstance.login(options),
    );
  };

  const handleSocialLogin = (connectionName: string) => {
    executeSafely(`Social login with connection: ${connectionName}`, () =>
      loginIdInstance.socialLogin({ connection: connectionName }),
    );
  };

  const handlePasskeyLogin = () => {
    const hasPasskeyData = !!loginIdInstance.screen?.data?.passkey;
    if (hasPasskeyData) {
      executeSafely(`Passkey login`, () => loginIdInstance.passkeyLogin());
    }
  };

  return {
    loginIdInstance,
    handleLoginId,
    handleSocialLogin,
    handlePasskeyLogin,
  };
};
