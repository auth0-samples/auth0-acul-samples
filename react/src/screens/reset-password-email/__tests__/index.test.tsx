import { render } from "@testing-library/react";

import { useResetPasswordEmailManager } from "../hooks/useResetPasswordEmailManager";
import ResetPasswordEmailScreen from "../index";

/// Mocking the hook
jest.mock("../hooks/useResetPasswordEmailManager", () => ({
  useResetPasswordEmailManager: jest.fn(),
}));

// Mocking the hooks and utility functions
jest.mock("../hooks/useResetPasswordEmailManager");

describe("ResetPasswordEmailScreen", () => {
  beforeEach(() => {
    // Clear all mocks before each test to ensure a clean slate
    jest.clearAllMocks();
  });

  it("should render correctly and match snapshot", () => {
    // Mock the hook's return values
    (useResetPasswordEmailManager as jest.Mock).mockReturnValue({
      resetPasswordEmail: {},
      texts: {
        pageTitle: "Password Reset",
        resendLinkText: "Resend verification email",
      },
      handleResendEmail: jest.fn(),
    });

    const { asFragment } = render(<ResetPasswordEmailScreen />);

    // Snapshot test to capture the rendered output
    expect(asFragment()).toMatchSnapshot();
  });

  it("should set the document title based on texts.pageTitle", () => {
    const mockPageTitle = "Password Reset";
    (useResetPasswordEmailManager as jest.Mock).mockReturnValue({
      resetPasswordEmail: {},
      texts: { pageTitle: mockPageTitle },
      handleResendEmail: jest.fn(),
    });

    render(<ResetPasswordEmailScreen />);

    // Assert that the document title matches the mocked value
    expect(document.title).toBe(mockPageTitle);
  });
});
