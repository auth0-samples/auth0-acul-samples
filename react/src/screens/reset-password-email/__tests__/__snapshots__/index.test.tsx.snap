// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ResetPasswordEmailScreen should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex px-10 py-20 justify-page-layout bg-background-page bg-(image:--ul-theme-page-bg-background-image-url) theme-universal"
  >
    <div
      class="bg-card text-card-foreground shadow-bevel-2xl flex flex-col rounded-4xl px-10 py-10 theme-universal:bg-widget-bg theme-universal:border-(--color-widget-border) theme-universal:rounded-widget theme-universal:shadow-widget theme-universal:border-(length:--border-widget) w-full max-w-[400px] gap-0"
      data-slot="card"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-mail flex flex-wrap justify-widget-logo ml-28"
        fill="none"
        height="80"
        stroke=""
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1"
        viewBox="0 0 24 24"
        width="80"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m22 7-8.991 5.727a2 2 0 0 1-2.009 0L2 7"
        />
        <rect
          height="16"
          rx="2"
          width="20"
          x="2"
          y="4"
        />
      </svg>
      <h1
        class="mt-6 mb-4 text-header justify-text-header text-(length:--ul-theme-font-title-size) font-title"
      >
        Check Your Email
      </h1>
      <p
        class="text-body-text justify-text-header text-(length:--ul-theme-font-subtitle-size) font-subtitle mb-6"
      >
        Please check the email address undefined for instructions to reset your password.
      </p>
      <button
        class="focus-visible:ring-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive theme-default:active:scale-[0.99] relative box-border inline-flex shrink-0 items-center justify-center gap-2 overflow-hidden text-sm font-medium whitespace-nowrap transition-all duration-150 ease-in-out outline-none focus-visible:ring-4 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 dark:bg-muted/50 hover:text-accent-foreground shadow-button-outlined-resting hover:shadow-button-outlined-hover hover:border-accent bg-background hover:bg-muted text-primary border-primary/35 theme-default:before:from-primary/5 theme-default:before:to-primary/0 theme-default:before:absolute theme-default:before:top-0 theme-default:before:left-0 theme-default:before:block theme-default:before:h-full theme-default:before:w-full theme-default:before:bg-gradient-to-t theme-default:before:content-[''] border h-10 rounded-2xl has-[>svg]:px-3 w-full p-6 cursor-pointer theme-universal:text-(--ul-theme-color-secondary-button-label) theme-universal:border-(--ul-theme-color-secondary-button-border) theme-universal:hover:shadow-[var(--button-hover-shadow)] theme-universal:border-1 theme-universal:focus:ring-4 theme-universal:focus:ring-base-focus/15 theme-universal:disabled:cursor-not-allowed theme-universal:rounded-button theme-universal:font-button theme-universal:text-(length:--ul-theme-font-buttons-text-size)"
        type="button"
      >
        Resend verification email
      </button>
    </div>
  </div>
</DocumentFragment>
`;
